pages/cookbook/jettons.mdx

```diff
@@ -1,9 +1,161 @@
+Note, that notifications are coming from YOUR contract's jetton wallet, so [`myAddress()`](/ref/core-common#myaddress) should be used in owner address field.
+Wallet initial data layout is shown below, but sometimes it can differ.
+
+```tact
+struct JettonWalletData {
+    balance: Int as coins;
+    ownerAddress: Address;
+    jettonMasterAddress: Address;
+    jettonWalletCode: Cell;
+}
+
+fun calculateJettonWalletAddress(ownerAddress: Address, jettonMasterAddress: Address, jettonWalletCode: Cell): Address {
+    let initData = JettonWalletData{
+        balance: 0,
+        ownerAddress,
+        jettonMasterAddress,
+        jettonWalletCode,
+    };
+
+    return contractAddress(StateInit{code: jettonWalletCode, data: initData.toCell()});
+}
+
+contract Sample {
+    jettonWalletCode: Cell;
+    jettonMasterAddress: Address;
+
+    init(jettonWalletCode: Cell, jettonMasterAddress: Address) {
+        self.jettonWalletCode = jettonWalletCode;
+        self.jettonMasterAddress = jettonMasterAddress;
+    }
+
+    receive(msg: TransferNotification) {
+        let myJettonWalletAddress = calculateJettonWalletAddress(myAddress(), self.jettonMasterAddress, self.jettonWalletCode);
+        require(sender() == myJettonWalletAddress, "Notification not from your jetton wallet!");
+
+        // your logic of processing token notification
+    }
+}
+```
+
+## Sending jetton transfer
+
+To send jetton transfer use [`send(){:tact}`](/book/send) function.
+If you want to have full control over the sending process and require some complex logic for it,
+use [`nativeSendMessage(){:tact}`](/ref/core-advanced#nativesendmessage) function.
+Note that `myJettonWalletAddress` may also be stored in contract storage to use less gas in every transaction.
+
+```tact
+message(0xf8a7ea5) TokenTransfer {
+    queryId: Int as uint64;
+    amount: Int as coins;
+    destination: Address;
+    excessesDestination: Address;
+    customPayload: Cell? = null;
+    forwardTonAmount: Int as coins;
+    forwardPayload: Slice as remaining;
+}
+
+receive("send") {
+    let myJettonWalletAddress = calculateJettonWalletAddress(myAddress(), self.jettonMasterAddress, self.jettonWalletCode);
+    send(SendParameters{
+        to: myJettonWalletAddress,
+        value: ton("0.05"),
+        body: TokenTransfer{
+            queryId: 42,
+            amount: jettonAmount, // jetton amount you want to transfer
+            destination: msg.userAddress, // address you want to transfer jettons. Note that this is address of jetton wallet owner, not jetton wallet itself
+            excessesDestination: msg.userAddress, // address were left toncoins will go
+            customPayload: null, // in most cases will be null and can be omited. Needed for custom logic on jettons itself
+            forwardTonAmount: 1, // amount that will be transferred with TransferNotification. Needed for custom logic execution like in example below
```

Comment:
we should also mention that the notification won't be sent if this amount is `0`.

Reply:
resolved